### PR TITLE
add statistics api for projects and repositories

### PR DESCRIPTION
**Pull request by wemeya**

### CODE REVIEW

**Comment by reasonerjt** [comment]
@wemeya could you let us know details of the API?

File: api/statistic.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+//GetReposByProject returns repo numbers of specified project
+func (s *StatisticAPI) GetReposByProject(projectName string, isAdmin bool) int {
+	repoList, err := svc_utils.GetRepoFromCache()
+	if err != nil {
+		log.Errorf("Failed to get repo from cache, error: %v", err)
+		s.RenderError(http.StatusInternalServerError, "internal sever error")
+	}
+	if isAdmin {
+		return len(repoList)
```
Comment (ywk253100) [inline api/statistic.go]: The value of len(repoList) is the count of all  repositories, not only for one project.

File: api/project.go [hunk 2]
```diff
@@ -126,10 +126,12 @@ func (p *ProjectAPI) Get() {
 		log.Errorf("Error occurred in QueryProject, error: %v", err)
 		p.CustomAbort(http.StatusInternalServerError, "Internal error.")
 	}
+	sa := &StatisticAPI{userID: p.userID, username: ""}
 	for i := 0; i < len(projectList); i++ {
-		if isProjectAdmin(p.userID, projectList[i].ProjectID) {
+		if projectList[i].Role == 1 {
```
Comment (ywk253100) [inline api/project.go]: Use models.PROJECTADMIN instead of 1.

File: dao/project.go [hunk 3]
```diff
@@ -84,19 +84,19 @@ func QueryProject(query models.Project) ([]models.Project, error) {
...
 	sql := `select distinct
-		p.project_id, p.owner_id, p.name,p.creation_time, p.update_time, p.public 
-	 from project p 
-		left join project_member pm on p.project_id = pm.project_id
-	 where p.deleted = 0 `
-
+		p.project_id, p.owner_id, p.name,p.creation_time, p.update_time, p.public`
 	queryParam := make([]interface{}, 1)
-
+	isAdmin, _ := IsAdminRole(query.UserID)
```
Comment (ywk253100) [inline dao/project.go]: Handle error.

File: models/project.go [hunk 4]
```diff
@@ -34,4 +34,6 @@ type Project struct {
 	Togglable bool
 
 	UpdateTime time.Time `orm:"update_time" json:"update_time"`
+	Role       int       `json:"RoleId"`
```
Comment (ywk253100) [inline models/project.go]: `json:"role_id"`

File: models/project.go [hunk 5]
```diff
@@ -34,4 +34,6 @@ type Project struct {
...
+	Repos      int
```
Comment (ywk253100) [inline models/project.go]: Repos ----> RepoCount. And add json description for it.

File: api/statistic.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+		if isAdmin {
+			proMap["total_repos"] = s.GetReposByProject(projectList[i].Name, true)
+		}
+	}
+	s.Data["json"] = proMap
+	s.ServeJSON()
+}
+
+//GetReposByProject returns repo numbers of specified project
+func (s *StatisticAPI) GetReposByProject(projectName string, isAdmin bool) int {
```
Comment (ywk253100) [inline api/statistic.go]: Make this function as a util func.

File: api/statistic.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	if err0 != nil {
+		log.Errorf("Error occured in check admin, error: %v", err0)
+		s.CustomAbort(http.StatusInternalServerError, "Internal error.")
+	}
+	if isAdmin {
+		proMap["total_project_count"] = len(projectList)
+		proMap["total_repo_count"] = getTotalRepoCount()
+	}
+	for i := 0; i < len(projectList); i++ {
+		if projectList[i].Role == models.PROJECTADMIN || projectList[i].Role == models.DEVELOPER {
```
Comment (ywk253100) [inline api/statistic.go]: The count should also +1 when the role is guest.

File: api/statistic.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+//getReposByProject returns repo numbers of specified project
+func getRepoCountByProject(projectName string) int {
+	repoList, err := svc_utils.GetRepoFromCache()
+	if err != nil {
+		log.Errorf("Failed to get repo from cache, error: %v", err)
+		return 0
+	}
+	var resp []string
```
Comment (ywk253100) [inline api/statistic.go]: All you need is a count, so you can just declare an int variable.

File: api/statistic.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+	isAdmin, err := dao.IsAdminRole(s.userID)
+	if err != nil {
+		log.Errorf("Error occured in check admin, error: %v", err)
+		s.CustomAbort(http.StatusInternalServerError, "Internal error.")
+	}
+	if isAdmin {
+		proMap["total_project_count"] = len(projectList)
+		proMap["total_repo_count"] = getTotalRepoCount()
+	}
+	for i := 0; i < len(projectList); i++ {
```
Comment (ywk253100) [inline api/statistic.go]: It seems that when the length of projectList is 0, the response is not correct.

Comment (wemeya) [inline api/statistic.go]: I'll handle that case.

File: dao/project.go [hunk 10]
```diff
@@ -107,11 +113,16 @@ func QueryProject(query models.Project) ([]models.Project, error) {
 	sql += " order by p.name "
 
 	var r []models.Project
-	_, err := o.Raw(sql, queryParam).QueryRows(&r)
+	_, err0 := o.Raw(sql, queryParam).QueryRows(&r)
```
Comment (ywk253100) [inline dao/project.go]: Use err instead of err0.

Comment (wemeya) [inline dao/project.go]: @ywk253100 OK.

**Comment by ywk253100** [comment]
LGTM

**Comment by ywk253100** [comment]
LGTM

File: dao/project.go [hunk 11]
```diff
@@ -228,6 +192,38 @@ func GetUserRelevantProjects(userID int) ([]models.Project, error) {
 	return projects, nil
 }
 
+// GetUserRelevantProjects returns the projects based on publicity and user, disregarding the names etc.
+func GetUserRelevantProjects(query models.Project) ([]models.Project, error) {
```
Comment (ywk253100) [inline dao/project.go]: The function should be split into two func: GetUserRelevantProjects and GetPublicProjects.

File: dao/project.go [hunk 12]
```diff
@@ -228,12 +192,61 @@ func GetUserRelevantProjects(userID int) ([]models.Project, error) {
 	return projects, nil
 }
 
+// GetUserRelevantProjects returns the projects based on publicity and user, disregarding the names etc.
```
Comment (ywk253100) [inline dao/project.go]: Modify the comment of function.

File: dao/project.go [hunk 13]
```diff
@@ -228,12 +192,61 @@ func GetUserRelevantProjects(userID int) ([]models.Project, error) {
...
+		return nil, err
+	}
+	return r, nil
+}
+
+//GetPublicProjects return all public projects whose name like projectName
+func GetPublicProjects(projectName string) ([]models.Project, error) {
+	o := orm.NewOrm()
+	var publicProjects []models.Project
+	sql := `select project_id, name, public
```
Comment (ywk253100) [inline dao/project.go]: Are these fields in sql enough?

File: dao/project.go (hunk 12 above)
Comment (wemeya) [inline dao/project.go]: @ywk253100  Ok, I will change that.

File: dao/project.go (hunk 13 above)
Comment (wemeya) [inline dao/project.go]: @ywk253100 I have not noticed that, it seems that's enough, I will check and add some fields if necessary.

File: dao/project.go [hunk 14]
```diff
@@ -228,12 +192,61 @@ func GetUserRelevantProjects(userID int) ([]models.Project, error) {
...
+
+//GetPublicProjects returns all public projects whose name like projectName
+func GetPublicProjects(projectName string) ([]models.Project, error) {
+	o := orm.NewOrm()
+	var publicProjects []models.Project
+	sql := `select project_id, owner_id, creation_time, update_time, name, public 
+		from project
+		where deleted = 0 and public=1`
+	if len(projectName) > 0 {
+		sql += " and name like '" + projectName + "'"
```
Comment (ywk253100) [inline dao/project.go]: SQL injection risk

File: api/project.go [hunk 15]
```diff
@@ -113,23 +113,50 @@ func (p *ProjectAPI) Head() {
...
+			projectList, err = dao.GetUserRelevantProjects(p.userID, projectName)
+		}
 	}
-	public, _ := p.GetInt("is_public")
-	queryProject.Public = public
-
-	projectList, err := dao.QueryProject(queryProject)
 	if err != nil {
-		log.Errorf("Error occurred in QueryProject, error: %v", err)
+		log.Errorf("Error occured in GetUserRelevantProjects, error: %v", err)
```
Comment (ywk253100) [inline api/project.go]: Modify the log info.

**Comment by ywk253100** [comment]
LGTM

**Comment by wemeya** [comment]
@ywk253100  Thanks a lot!

**Comment by reasonerjt** [comment]
Seems GetAllProjects and GetPublicProjects share a lot code, why not merge them into one? You call create a private func and let GetAllProjects GetPublicProjects call it with different parms?

**Comment by wemeya** [comment]
@reasonerjt Good advice! I've modified as you said, thanks a lot!

**Pull request merged**